### ISSUE TITLE
twisted.test.test_internet calls reactor.iterate()

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10798]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#1200|
|Type|defect|
|Created|2005-09-19 18:01:03Z|
|Branch|https://github.com/twisted/twisted/tree/reactor-iterate-1200|

Attachments:

* [remove-uses-of-reactor-iterate](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/73e/73ee4958bdb5a056029ebd39b8abbaa3dbc0f333/d8d7972e2b42c4b668b361722dd7abd7d6a3eee9) (7048 bytes) - added by danderson on 2009-09-13 05:45:32Z - Patch for this issue.
* [remove-uses-of-reactor-iterate2](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/73e/73ee4958bdb5a056029ebd39b8abbaa3dbc0f333/20d194f19a6fab0f7286e71cd53e94e47a8d7fbe) (7697 bytes) - added by danderson on 2009-09-15 00:42:33Z - Patch for this issue.

<details><summary>Searchable metadata</summary>

```
trac-id__1200 1200
type__defect defect
reporter__exarkun exarkun
priority__high high
milestone__ 
branch__branches_reactor_iterate_1200 branches/reactor-iterate-1200
branch_author__danderson__thijs danderson, thijs
status__closed closed
resolution__fixed fixed
component__core core
keywords__tests_easy tests easy
time__1127152863000000 1127152863000000
changetime__1258037332000000 1258037332000000
version__ 
owner__ 
cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
Some of these are necessary, I guess.  Not all of them, though.  The ones near
threading look like they can probably go, at least.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

Remaining calls to `reactor.iterate` in `test_internet.py`:

  1. three in `testCallInNextIteration`.  This test method has no docstring and was apparently introduced in a passing state (r11909).  Sometimes gtk2reactor (or maybe another glib-related reactor) fails this test.  As far as I can tell, this is just a test for the interaction between `reactor.iterate` and `reactor.callLater`, nothing more subtle.  It could:
    1. be deleted
    1. move to a reactor.iterate-specific test case
  1. one in `ReactorCoreTestCase.test_iterate`.  This seems to be a test for `reactor.iterate` directly.  It should be moved to `twisted.internet.test.test_core` where it can use a non-deprecated version of `reactor.iterate`.
  1. one in `DelayedTestCase.setUp`, along with some other really nasty crap.  This seems to be code for cleaning up the reactor after some other tests have left it dirty.  This should not be necessary any more.  It should be deleted.
  1. one in `DelayedTestCase.testActive` which is testing that after a delayed call fires, its `active` method returns a false value.  It should be rewritten to not use `reactor.iterate`, it should be trivial.
  1. one in a loop in `CallFromThreadTestCase.testScheduling`, this seems to be making sure that the reactor processes threading events.  It could be replaced with something event-driven.
  1. three in `CallFromThreadTestCase.testCorrectOrder`, similar to the previous usage, should be replaceable with something event driven.
  1. one in `CallFromThreadTestCase.testNotRunAtOnce`, similar to the previous usage, should be replaceable with something event driven.

The first point might be tricky.  The rest should be straightforward.  If so, perhaps the first point can be split off into a separate ticket, so as not to block the rest of these fixes.

**Comment by twisted-trac** [comment]
|<img alt="danderson's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|danderson removed owner|
|-|-|

Replying to [exarkun](#note_2):
>   1. three in `testCallInNextIteration`.  This test method has no docstring and was apparently introduced in a passing state ().  Sometimes gtk2reactor (or maybe another glib-related reactor) fails this test.  As far as I can tell, this is just a test for the interaction between `reactor.iterate` and `reactor.callLater`, nothing more subtle.  It could:
>     1. be deleted
>     1. move to a reactor.iterate-specific test case

Deleted. As far as I can tell, it's testing a behavior that isn't terribly interesting, and assumes that a reactor's iteration will not run anything scheduled in that iteration. I don't see why that should particularly be a guarantee of IReactorCore, as long as iterate() doesn't block as run() does and returns fairly quickly.

>   1. one in `ReactorCoreTestCase.test_iterate`.  This seems to be a test for `reactor.iterate` directly.  It should be moved to `twisted.internet.test.test_core` where it can use a non-deprecated version of `reactor.iterate`.

Moved as suggested, and `ReactorCoreTestCase` deleted from test_internet.py, since it was duplicating tests that already exist (and are more comprehensive) in `twisted.internet.test.test_core`.

>   1. one in `DelayedTestCase.setUp`, along with some other really nasty crap.  This seems to be code for cleaning up the reactor after some other tests have left it dirty.  This should not be necessary any more.  It should be deleted.

Crap deleted. Tests continue to pass on my machine.

>   1. one in `DelayedTestCase.testActive` which is testing that after a delayed call fires, its `active` method returns a false value.  It should be rewritten to not use `reactor.iterate`, it should be trivial.

Replaced with an event-driven test.

>   1. one in a loop in `CallFromThreadTestCase.testScheduling`, this seems to be making sure that the reactor processes threading events.  It could be replaced with something event-driven.

Rewritten.

>   1. three in `CallFromThreadTestCase.testCorrectOrder`, similar to the previous usage, should be replaceable with something event driven.

Rewritten.

>   1. one in `CallFromThreadTestCase.testNotRunAtOnce`, similar to the previous usage, should be replaceable with something event driven.

Rewritten.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to danderson|
|-|-|

Thanks for your work on this danderson.  Lots of older tickets like this one rarely get attention.

  1. `IReactorCore.iterate` is certainly underspecified.  `testCallInNextIteration` seems to have been added in an already-passing state, so it's not really clear what it was intended to demonstrate.  This test sometimes fails on the Gtk or Gtk2 reactor, which may not be terribly interesting (at least inasmuch as `iterate` itself is not all that interesting)`.  The behavior of `reactor.callLater(0, ...)` waiting until some I/O has been processed before the scheduled function is called '''is''' interesting, though.  If reactors don't implement this, then it becomes somewhat trivial to block any I/O from being processed with an apparently safe scheduler loop.  So this test should probably be reformulated as something which does not use `iterate` but which does verify that timing and i/o events interleave nicely.  This can be another ticket, and just deleting the test from the suite for this ticket is fine, since this test doesn't actually exercise this behavior.
  1. A potential source of problems that `ReactorCoreTestCase.test_iterate` had was its reliance on `time.time` to determine whether or not the right thing was happening.  Looking at old build data, it doesn't seem like this ever actually caused the test to fail spuriously, so it's probably okay to move it more or less as-is.  However, the change from using 8 to 2 is probably not merited without careful consideration and explanation.
  1. `DelayedTestCase.testActive` should be renamed to `test_active`, its docstring should be reformatted to match the coding style guide, and it would probably benefit from using `def` instead of `lambda` to define the callback that does the assertion.
  1. Likewise `CallFromThreadTestCase.testScheduling`, `testCorrectOrder`, and `testNotRunAtOnce` should be renamed and given properly formatted docstrings describing what they cover.
  1. `testCorrectOrder` used to make sure each call happened in the right order.  Now it only looks at whether the last call happens last.  Also, it won't actually fail if the code doesn't work right, since all it does is fire the test `Deferred` with `False` if the expectation is not met.  A `Deferred` that fires with a `Failure` fails a test, but firing with `False` does not.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="danderson's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|danderson commented|
|-|-|

Replying to [exarkun](#note_4):
>   1. `IReactorCore.iterate` is certainly underspecified.  `testCallInNextIteration` seems to have been added in an already-passing state, so it's not really clear what it was intended to demonstrate.  This test sometimes fails on the Gtk or Gtk2 reactor, which may not be terribly interesting (at least inasmuch as `iterate` itself is not all that interesting)`.  The behavior of `reactor.callLater(0, ...)` waiting until some I/O has been processed before the scheduled function is called '''is''' interesting, though.  If reactors don't implement this, then it becomes somewhat trivial to block any I/O from being processed with an apparently safe scheduler loop.  So this test should probably be reformulated as something which does not use `iterate` but which does verify that timing and i/o events interleave nicely.  This can be another ticket, and just deleting the test from the suite for this ticket is fine, since this test doesn't actually exercise this behavior.

Good point. Opened #4026 to track a correct reimplementation of this test.

>   1. A potential source of problems that `ReactorCoreTestCase.test_iterate` had was its reliance on `time.time` to determine whether or not the right thing was happening.  Looking at old build data, it doesn't seem like this ever actually caused the test to fail spuriously, so it's probably okay to move it more or less as-is.  However, the change from using 8 to 2 is probably not merited without careful consideration and explanation.

Given the structure of the test and lack of external I/O, I expect this test to complete in well under half a second on the slowest of slow machines. It should be basically just blazing through the reactor code, not doing anything at all. I can raise the limit if you'd like, but I'm willing to bet a beer that nothing dire will happen with the current limit.

>   1. `DelayedTestCase.testActive` should be renamed to `test_active`, its docstring should be reformatted to match the coding style guide, and it would probably benefit from using `def` instead of `lambda` to define the callback that does the assertion.

Done.

>   1. Likewise `CallFromThreadTestCase.testScheduling`, `testCorrectOrder`, and `testNotRunAtOnce` should be renamed and given properly formatted docstrings describing what they cover.

All done.

>   1. `testCorrectOrder` used to make sure each call happened in the right order.  Now it only looks at whether the last call happens last.  Also, it won't actually fail if the code doesn't work right, since all it does is fire the test `Deferred` with `False` if the expectation is not met.  A `Deferred` that fires with a `Failure` fails a test, but firing with `False` does not.

Changed so that the final thread in sequence fires the test Deferred in the reactor thread. The test now happens in the Deferred's callback chain, which correctly propagates failure exceptions back up to trial.

There is another test using reactor.iterate in test_internet, `TimeTestCase.testCallLaterDelayAndReset`. However, untangling reactor.iterate from there is more involved, since its use is hidden in a clock test helper. Opened [#4027](https://github.com/twisted/twisted/issues/4027) to deal with that.

I'll be AFK until Friday, so if there are more tweaks to do on this bug, don't panic if I don't react until then.

**Comment by twisted-trac** [comment]
|<img alt="danderson's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|danderson removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_7"></a>|@jesstess commented|
|-|-|

Just a tiny nit that the docstring quotes for test_iterate should be on lines by themselves.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_8"></a>|@thijstriemstra commented|
|-|-|

(In [27423]) Branching to 'reactor-iterate-1200'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_9"></a>|@thijstriemstra commented|
|-|-|

(In [27425]) Docstring fixes, refs [#10798](https://github.com/twisted/twisted/issues/10798)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_10"></a>|@thijstriemstra commented|
|-|-|

Applied the patch in a branch, up for review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

Sorry for the long turn-around on this review.  A few minor points which I will correct, after which I will merge the branch:

  1. `test_lotsOfThreadsAreScheduledCorrectly`'s docstring is misleading, it suggests threads are being used, but `callFromThread` is the opposite - it's the API that runs a function in the _reactor_ thread.
  1. There should be two blank lines between methods, three blank lines between classes.
  1. Method names in docstrings should be marked up with `L{}`.
  1. The `threading` import in `test_internet.py` is unused

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set status to `closed`|
|-|-|

(In [27455]) Merge reactor-iterate-1200

Author: danderson, thijs
Reviewer: exarkun
Fixes: [#10798](https://github.com/twisted/twisted/issues/10798)
Refs: [#4027](https://github.com/twisted/twisted/issues/4027)

Remove uses of `reactor.iterate` from `twisted.test.test_internet`.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
